sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,346 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans.logical.estimation
+
+import scala.collection.immutable.{HashSet, Map}
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.util.DateTimeUtils
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.UTF8String
+
+
+object FilterEstimation extends Logging {
+
+  // We use a mutable colStats because we need to update the corresponding ColumnStat
+  // for a column after we apply a predicate condition.
+  private val mutableColStats: mutable.Map[String, ColumnStat] = mutable.Map.empty[String, ColumnStat]
+
+  def estimate(plan: Filter)
+    : Option[Statistics] = {
+    val stats: Statistics = plan.child.statistics
+    if (stats.rowCount.isEmpty) return None
+
+    copyFromColStats(stats.colStats)
+    // estimate selectivity for this filter
+    val percent: Double = calculateConditions(stats, plan.condition)
+    val filteredSizeInBytes = RoundingToBigInt(BigDecimal(stats.sizeInBytes) * percent)
+    val filteredRowCount = stats.rowCount.map(
+      r => RoundingToBigInt(BigDecimal(r) * percent)
+    )
+
+    Some(stats.copy(sizeInBytes = filteredSizeInBytes, rowCount = filteredRowCount,
+      colStats = copyToColStats))
+  }
+
+  // save local copy of colStats so that we may update it
+  def copyFromColStats(colStats: Map[String, ColumnStat]): Unit = {
+    for ( (k, v) <- colStats ) {
+      mutableColStats += (k -> v)
+    }
+  }
+  // return the updated colStats to the calling method.
+  def copyToColStats: Map[String, ColumnStat] = {
+    var stats: Map[String, ColumnStat] = Map.empty[String, ColumnStat]
+    for ( (k, v) <- mutableColStats ) {
+      stats = stats + (k -> v)
+    }
+    stats
+  }
+
+  def calculateConditions(
+      planStat: Statistics,
+      condition: Expression,
+      update: Boolean = true)
+    : Double = {
+    // For conditions linked by And, we need to update stats after a condition estimation
+    // so that the stats will be more accurate for subsequent estimation.
+    condition match {
+      case And(cond1, cond2) =>
+        calculateConditions(planStat, cond1, update) * calculateConditions(planStat, cond2, update)
+      case Or(cond1, cond2) =>
+        math.min(1.0, calculateConditions(planStat, cond1, update = false) +

```

Comment:
Sorry, I meant min(c1,c2), how about this? I think c1+c2 is too loose, and if we really want to do it this way, it should be c1+c2- (c1 and c2)